### PR TITLE
112 pf preprocessing (Comparing Pre-processing changes)

### PR DESCRIPTION
**Pull request by p-flock**
@KinanBab let's go through this together

### CODE REVIEW

**Comment by p-flock** [comment]
after your talk

File: lib/jiff-client.js [hunk 1]
```diff
@@ -4164,27 +4381,62 @@
...
         compute_threshold = compute_list.length;
       }
 
-      var bit = jiff.helpers.random(2);
-      var bit_shares = jiff.internal_share(bit, compute_threshold, compute_list, compute_list, Zp, op_id + ':share');
+      var random_bit;
+      if (compute_list.indexOf(jiff.id) > -1) {
+        var bit = jiff.helpers.random(2);
+        // reduce the threshold to make bgw possible -> won't support n-1 security
+        var bit_shares = jiff.internal_share(bit, compute_threshold/2, compute_list, compute_list, Zp, op_id + ':share');
```
Comment (p-flock) [inline lib/jiff-client.js]: should use Math.floor((compute_threshold+1)/2) for second argument.

File: lib/jiff-client.js [hunk 2]
```diff
@@ -4164,27 +4381,62 @@
...
+     *                         ensure that corresponding instructions across different parties are matched correctly.
+     * @param {object} [params={bit_length}] - an object containing extra parameters passed by the user.
+     *                               Expects:
+     *                               - count: how many random bits should be computed
+     *                               - an optional number compute_threshold parameter, which specifies threshold used
+     *                               during the protocol execution. By default, this is the length of the compute_list.
+     * @return {[SecretShare]} An array with this party's share of each generated bit
+     */
+    jiff.protocols.generate_bit_sequence = function (threshold, receivers_list, compute_list, Zp, op_id, params) {
+      if (params.count == null) {
```
Comment (p-flock) [inline lib/jiff-client.js]: if(params == null) ...

File: lib/jiff-client.js [hunk 3]
```diff
@@ -4245,7 +4497,8 @@
       var d = [];
       d[c.length - 1] = c[c.length - 1];
       for (i = c.length - 2; i >= 0; i--) {
-        d[i] = d[i + 1].isor_bit(c[i], op_id + ':sOR:' + i);
+        d[i] = d[i + 1].isor_bit(c[i], op_id + ':sOR:' + i); //does the op_id need to increment?
+        //other option is rely on sor_bit and find a way to preprocess for a dynamic number of sor_bit operations
```
Comment (p-flock) [inline lib/jiff-client.js]: delete the comment plz

File: lib/jiff-client.js [hunk 4]
```diff
@@ -4420,21 +4673,240 @@
...
+    jiff.preprocessing_function_map['clteq'] = jiff.preprocessing_function_map['cgt'];
+    jiff.preprocessing_function_map['cgteq'] = jiff.preprocessing_function_map['clt'];
+    // equality / constant equality
+    jiff.preprocessing_function_map['seq'] = jiff.preprocessing_function_map['clteq'];
+    jiff.preprocessing_function_map['sneq'] = jiff.preprocessing_function_map['seq'];
+    jiff.preprocessing_function_map['ceq'] = jiff.preprocessing_function_map['clteq'];
+    jiff.preprocessing_function_map['cneq'] = jiff.preprocessing_function_map['ceq'];
+
+    jiff.preprocessing_function_map['or'] = jiff.preprocessing_function_map['sor']; // would not worry about secret/constant neutral versions for now
+    jiff.preprocessing_function_map['xor'] = jiff.preprocessing_function_map['sxor'];
```
Comment (p-flock) [inline lib/jiff-client.js]: remove the non s or c prefixed operations

File: lib/jiff-client.js [hunk 5]
```diff
@@ -4420,21 +4673,240 @@
...
-    jiff.counters.number_op_count = {};
-    jiff.counters.op_count = {};
-    jiff.counters.share_obj_count = 0;
-    jiff.counters.pending_opens = 0; // Keeps track of pending opens to disconnect safely.
+    jiff.counters.reset = function () {
+      jiff.counters.share_op_count = {};
+      jiff.counters.open_op_count = {};
+      jiff.counters.triplet_op_count = {};
+      jiff.counters.number_op_count = {};
+      jiff.counters.op_count = {};
```
Comment (p-flock) [inline lib/jiff-client.js]: Add op counters per primitive, and simplify pre-processing

Comment (p-flock) [reply lib/jiff-client.js]: will entail chaning gen_op_id,
note: this is sparse, do not store zeros

File: lib/jiff-client.js [hunk 6]
```diff
@@ -4420,21 +4673,240 @@
...
+        count = 1;
+      }
+
+      // default protocols
+      if (protocols == null) {
+        protocols = {};
+      }
+      if (protocols.beaver == null) {
+        protocols.beaver = jiff.protocols.generate_beaver_bgw;
+        protocols.generate_bit_sequence = jiff.protocols.generate_bit_sequence;
```
Comment (p-flock) [inline lib/jiff-client.js]: separate into two if statements (looking at using && short circuited)

File: lib/jiff-client.js [hunk 7]
```diff
@@ -4420,21 +4673,240 @@
...
+     *                               the function must implement the same interface as the JIFF provided protocols (e.g. jiff.protocols.generate_beaver_bgw).
+     *                               missing mappings indicate that JIFF must use the default protocols.
+     * @param {Number} [threshold=receivers_list.length] - the threshold of the preprocessed shares.
+     * @param {Array} [receivers_list=all_parties] - the parties that will receive the preprocsssed shares.
+     * @param {Array} [compute_list=all_parties] - the parties that will compute the preprocsssed shares.
+     * @param {Array} [Zp=jiff.Zp] - the Zp of the preprocessed shares.
+     * @param {Array} [id_list=auto_gen()] - array of ids to be used sequentially to identify the pre_processed values. Optional.
+     * @param {Object} [params={}] - any additional protocol-specific parameters.
+     */
+    jiff.preprocessing = function (dependent_op, count, protocols, threshold, receivers_list, compute_list, Zp, id_list, params) {
```
Comment (p-flock) [inline lib/jiff-client.js]: add if statement for invalid dependent_op (null, does not exist in protocols, does not exist in pre-processing map)

File: lib/jiff-client.js [hunk 8]
```diff
@@ -4420,21 +4673,240 @@
...
+        { op:'beaver', op_id: ':triplet' }
+      ],
+      'sxor_bit' : [
+        { op:'smult', op_id:''} // no op_id suffix according to sxor_bit implementation
+      ],
+      'sor_bit' : [
+        { op:'smult', op_id:':smult1'}
+      ],
+      'lt_halfprime' : [
+        { op:'generate_bit_sequence', op_id:':bitsequence', params: {
```
Comment (p-flock) [inline lib/jiff-client.js]: indentation

File: lib/jiff-client.js [hunk 9]
```diff
@@ -4420,21 +4673,240 @@
...
+          jiff.store_preprocessing(id, result.share);
+          promises.push(result.promise);
+        } else {
+          for (var k = 0; k < preprocessing_dependencies.length; k++) {
+            var dependency = preprocessing_dependencies[k];
+
+            var next_op = dependency['op'];
+            // copy both the originally given extra_params and the extra params of the dependency and merge them
+            // together, dependency params overwrite duplicate keys.
+            var extra_params = Object.assign({}, params, dependency['params']);
```
Comment (p-flock) [inline lib/jiff-client.js]: add comment in the code to change this, if we ever add params to non leafs

**Comment by KinanBab** [comment]
TODO:
* Pre-processing for extensions and helpers (if_else)
* Make sure tests are stable (batching pre-processing / barriers)
* Small Cleanups
* Op_ids per operation
* Make count of certain operations dynamic (sdiv, comparisons)
* Pre-processing in pre-processing?
* Pre-processing store and get Hooks
* Flag for reverting to server in case preprocessing did not exist

**Pull request closed without merging**